Source/Private/ASCollectionLayout.mm

```diff
@@ -87,70 +85,77 @@ - (void)prepareLayout
   [super prepareLayout];
   ASCollectionLayoutContext *context = [self layoutContextWithElements:_collectionNode.visibleElements];
   
-  ASCollectionLayoutState *state = nil;
+  ASCollectionLayoutState *layout = nil;
   {
     ASDN::MutexLocker l(__instanceLock__);
-    if (_pendingState != nil && ASObjectIsEqual(_layoutContextForPendingState, context)) {
-      // Looks like we can use the pending state. Great!
-      state = _pendingState;
-      _pendingState = nil;
-      _layoutContextForPendingState = nil;
+    if (_pendingLayout != nil && ASObjectIsEqual(_pendingLayout.context, context)) {
+      // Looks like we can use the pending layout. Great!
+      layout = _pendingLayout;
+      _pendingLayout = nil;
     }
   }
   
-  if (state == nil) {
-    state = [_layoutDelegate calculateLayoutWithContext:context];
+  if (layout == nil) {
+    layout = [_layoutDelegate calculateLayoutWithContext:context];
   }
   
-  _state = state;
+  _layout = layout;
 }
 
 - (void)invalidateLayout
 {
   ASDisplayNodeAssertMainThread();
   [super invalidateLayout];
-  _state = nil;
+  _layout = nil;
 }
 
 - (CGSize)collectionViewContentSize
 {
   ASDisplayNodeAssertMainThread();
-  ASDisplayNodeAssertNotNil(_state, @"Collection layout state should not be nil at this point");
-  return _state.contentSize;
+  ASDisplayNodeAssertNotNil(_layout, @"Collection layout state should not be nil at this point");
+  return _layout.contentSize;
 }
 
 - (NSArray<UICollectionViewLayoutAttributes *> *)layoutAttributesForElementsInRect:(CGRect)rect
 {
-  NSMutableArray *attributesInRect = [NSMutableArray array];
-  NSMapTable *attrsMap = _state.elementToLayoutArrtibutesMap;
-  for (ASCollectionElement *element in attrsMap) {
-    UICollectionViewLayoutAttributes *attrs = [attrsMap objectForKey:element];
-    if (CGRectIntersectsRect(rect, attrs.frame)) {
-      [ASCollectionLayout setSize:attrs.frame.size toElement:element];
-      [attributesInRect addObject:attrs];
-    }
+  ASDisplayNodeAssertMainThread();
+  if (CGRectIsEmpty(rect) || (! CGRectIntersectsRect(_layout.contentRect, rect))) {
+    return @[];
+  }
+  
+  ASCollectionLayoutState *layout = _layout;
```

Comment:
No need to create these local copies of strong ivars – since the ivar isn't marked `volatile` the compiler is free to cache the value in a register. If there's gonna be a lot of underscores, this can make things prettier, and if the ivar is weak, then we definitely ought to get a strong local copy.